ably/helpers_test.go

```diff
@@ -18,6 +18,93 @@ func (opts *ClientOptions) RealtimeURL() string {
 	return opts.realtimeURL()
 }
 
+// StateRecorder provides:
+//
+//   * send ably.State channel for recording state transitions
+//   * goroutine-safe access to recorded state enums
+//
+type StateRecorder struct {
+	ch     chan State
+	states []StateEnum
+	wg     sync.WaitGroup
+	done   chan struct{}
+	mtx    sync.Mutex
+}
+
+func NewStateRecorder() *StateRecorder {
+	return newRecorderBuffer(16)
+}
+
+func newRecorderBuffer(buffer int) *StateRecorder {
+	rec := &StateRecorder{
+		ch:     make(chan State, buffer),
+		done:   make(chan struct{}),
+		states: make([]StateEnum, 0, 16),
+	}
+	rec.wg.Add(1)
+	go rec.processIncomingStates()
+	return rec
+}
+
+func (rec *StateRecorder) processIncomingStates() {
+	defer rec.wg.Done()
+	for {
+		select {
+		case state, ok := <-rec.ch:
+			if !ok {
+				return
+			}
+			rec.add(state.State)
+		case <-rec.done:
+			return
+		}
+	}
+}
+
+// Add appends state to the list of recorded ones, used to ensure ordering
+// of the states by injecting values at certain points of the test.
+func (rec *StateRecorder) Add(state StateEnum) {
+	rec.ch <- State{State: state}
+}
+
+func (rec *StateRecorder) add(state StateEnum) {
+	rec.mtx.Lock()
+	rec.states = append(rec.states, state)
+	rec.mtx.Unlock()
+}
+
+func (rec *StateRecorder) Channel() chan<- State {
+	return rec.ch
+}
+
+func (rec *StateRecorder) Options() *ClientOptions {
+	return &ClientOptions{Listener: rec.ch}
+}
+
+// Stop stops separate recording gorouting and waits until it terminates.
+func (rec *StateRecorder) Stop() {
+	close(rec.done)
+	rec.wg.Wait()
+	// Drain listener channel.
```

Comment:
@kouno No, we can't close it, the ownership of the channel is transferred to (*ClientOptions).Listener; the client may still send some events, like `StateConnClosing` etc., but we don't care about them since we stopped recording. We don't need to keep reading the channel because `ably.Client`'s event emitter sends state values in a non-blocking manner.